### PR TITLE
Add [buildkite] service

### PR DESCRIPTION
**Pull request by raulriera**
Fixes https://github.com/badges/shields/issues/1261 by providing Buildkite support for Shield.

Usage is:

- `/buildkite/{id}/{name}` for the default Buildkite shield `eg Buildkite | passing`

### CODE REVIEW

**Comment by raulriera** [comment]
Thanks for the comments, updated. But, I'm having some issues with the tests... The documentation to creating them is broken and I can't seem to find how to "run them". Doing `npm test` will run all the tests. But mine are always succeeding even when changing the expected values.

File: server.js [hunk 1]
```diff
@@ -6217,6 +6217,45 @@ cache(function(data, match, sendBadge, request) {
...
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+      return;
+    }
+
+    try {
+      var data = JSON.parse(buffer);
+      var status = data.status;
+      if (status === 'passing') {
+        badgeData.text[1] = 'Passing';
```
Comment (RedSparr0w) [inline server.js]: For the sake of consistency could you change these to be lowercase strings.
eg: travis-ci badge:
travis-build

File: server.js [hunk 2]
```diff
@@ -6217,6 +6217,45 @@ cache(function(data, match, sendBadge, request) {
...
 
+// Buildkite integration.
+camp.route(/^\/buildkite\/([^/]+)\/([^/]+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var identifier = match[1];  // eg, 3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489
+  var branch = match[2];  // eg, master
+  var format = match[3];
+
+  var url = 'https://badge.buildkite.com/' + identifier + '.json?branch=' + branch;
+  var badgeData = getBadgeData('buildkite', data);
```
Comment (RedSparr0w) [inline server.js]: For the sake of consistency could you change this to be `build`
eg: travis-ci badge:
travis-build

File: server.js [hunk 3]
```diff
@@ -6217,6 +6217,45 @@ cache(function(data, match, sendBadge, request) {
 }));
 
 
+// Buildkite integration.
+camp.route(/^\/buildkite\/([^/]+)\/([^/]+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var identifier = match[1];  // eg, 3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489
+  var branch = match[2];  // eg, master
```
Comment (RedSparr0w) [inline server.js]: Is `branch` required for this badge to work, maybe we could make this optional?
If we do require a branch to be specified we could possibly use something like this to make it optional:
```javascript
camp.route(/^\/buildkite\/([^/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/,
  ...
  const branch = match[2] || 'master'; // Defaults to master if not specified
```

File: services/buildkite/buildkite.tester.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+const t = new ServiceTester({ id: 'buildkite', title: 'Buildkite Builds' });
+module.exports = t;
+
+t.create('buildkite invalid pipeline')
+  .get('/unknown-identifier/unknown-branch.json')
+  .expectJSON({ name: 'buildkite', value: 'invalid' });
+
+t.create('buildkite valid pipeline')
+  .get('/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489/master.json')
+  .expectJSON({ name: 'buildkite', value: 'Passing' });
```
Comment (RedSparr0w) [inline services/buildkite/buildkite.tester.js]: Could you please also add the following tests:
- valid `identifier`/`id` with invalid `branch`
- [invalid JSON response](https://github.com/badges/shields/blob/master/services/cdnjs/cdnjs.tester.js#L28-L34)
- [connection error](https://github.com/badges/shields/blob/master/services/appveyor/appveyor.tester.js#L28-L31)

File: server.js [hunk 5]
```diff
@@ -6217,6 +6217,45 @@ cache(function(data, match, sendBadge, request) {
...
+  var format = match[3];
+
+  var url = 'https://badge.buildkite.com/' + identifier + '.json?branch=' + branch;
+  var badgeData = getBadgeData('buildkite', data);
+
+  request(url, function(err, res, buffer) {
+    if (err != null) {
```
Comment (RedSparr0w) [inline server.js]: Could you update this to use the [`checkErrorResponse()`](https://github.com/badges/shields/blob/master/lib/error-helper.js#L8-L24) helper here.
[Example usage](https://github.com/badges/shields/blob/master/server.js#L569-L572)

**Note:**
_if you want to change the default `not found` message on `404 not found` errors, you can add an extra (string) parameter to the end which will be returned instead_

File: lib/all-badge-examples.js [hunk 6]
```diff
@@ -155,8 +155,8 @@ const allBadgeExamples = [
         previewUri: '/appveyor/tests/NZSmartie/coap-net-iu0to/master.svg'
       },
       {
-        title: 'Codeship',
-        previewUri: '/codeship/d6c1ddd0-16a3-0132-5f85-2e35c05e22b1.svg'
```
Comment (RedSparr0w) [inline lib/all-badge-examples.js]: Looks like you may have inadvertently removed this Codeship example badge?

**Comment by RedSparr0w** [comment]
You can run just this services tests with:
```cmd
npm run test:services -- --only=buildkite
```

File: services/buildkite/buildkite.tester.js (hunk 4 above)
Comment (raulriera) [reply services/buildkite/buildkite.tester.js]: Having a hard time implementing the `invalid JSON response` one, I don't quite understand what is the linked example doing. Following something similar I get the same responses as connection error back 

```
 + expected - actual

       {
      -  "name": "build"
      -  "value": "inaccessible"
      +  "name": "buildkite"
      +  "value": "invalid"
       }
```

File: server.js (hunk 3 above)
Comment (raulriera) [reply server.js]: If not required, no. I'll change it

File: services/buildkite/buildkite.tester.js (hunk 4 above)
Comment (RedSparr0w) [reply services/buildkite/buildkite.tester.js]: [This line](https://github.com/badges/shields/blob/master/services/cdnjs/cdnjs.tester.js#L6) will also be needed for the `Invalid JSON response` test.
Strange that it's returning `inaccessible` though as it should only return that when there is an error response from `request()`

Comment (raulriera) [reply services/buildkite/buildkite.tester.js]: Should I be doing something different?

```
t.create('cdnjs (unexpected response)')
  .get('/_.json')
  .intercept(nock => nock('https://badge.buildkite.com')
    .get('/_.json')
    .reply(invalidJSON)
  )
  .expectJSON({name: 'buildkite', value: 'invalid'});
```

🤔

File: services/buildkite/buildkite.tester.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  .get('/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489/unknown-branch.json')
+  .expectJSON({ name: 'build', value: 'unknown' });
+
+t.create('buildkite connection error')
+  .get('/_.json')
+  .networkOff()
+  .expectJSON({ name: 'build', value: 'inaccessible' });
+
+t.create('buildkite unexpected response')
+  .get('/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489.json?branch=master')
```
Comment (RedSparr0w) [inline services/buildkite/buildkite.tester.js]: I think you may be able to leave out the `?branch=master` from this part of the request.

File: server.js [hunk 8]
```diff
@@ -6217,6 +6217,44 @@ cache(function(data, match, sendBadge, request) {
...
+      var status = data.status;
+      if (status === 'passing') {
+        badgeData.text[1] = 'passing';
+        badgeData.colorscheme = 'green';
+      } else if (status === 'failing') {
+        badgeData.text[1] = 'failing';
+        badgeData.colorscheme = 'red';
+      } else {
+        badgeData.text[1] = 'unknown';
+        badgeData.colorscheme = 'gray';
```
Comment (RedSparr0w) [inline server.js]: Last thing and then i think we are good for merge,
Could you update this to use `lightgray`.

`lightgray`:

`gray`:

File: services/buildkite/buildkite.tester.js (hunk 7 above)
Comment (raulriera) [reply services/buildkite/buildkite.tester.js]: How come? What’s makes it different from the one below?

Comment (RedSparr0w) [reply services/buildkite/buildkite.tester.js]: this 1st is the address you would enter to get the badges image.
eg:
```
https://img.shields.io/buildkite/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489.json?branch=master
```
would only need to be
```
https://img.shields.io/buildkite/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489.json
```
or
```
https://img.shields.io/buildkite/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489/master.json
```
--------------
the 2nd is the request we want to intercept to the buildkite server:
```
https://badge.buildkite.com/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489.json?branch=master
```

Comment (raulriera) [reply services/buildkite/buildkite.tester.js]: Ooohh I see. Thanks!👌

File: services/buildkite/buildkite.tester.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  .get('/unknown-identifier/unknown-branch.json')
+  .expectJSON({ name: 'build', value: 'not found' });
+
+t.create('buildkite valid pipeline')
+  .get('/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489/master.json')
+  .expectJSON({ name: 'build', value: 'passing' });
+
+t.create('buildkite valid pipeline skipping branch')
+  .get('/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489.json')
+  .expectJSON({ name: 'build', value: 'passing' });
```
Comment (RedSparr0w) [inline services/buildkite/buildkite.tester.js]: Sorry just noticed this also,
Could you update the relevant test to use:
```javascript
  .expectJSONTypes(Joi.object().keys({
    name: 'build',
    value: Joi.equal('failing', 'passing', 'unknown')
  }));
```
This just ensures our test won't fail if the service that we are testing against fails :smile:

**Comment by RedSparr0w** [comment]
Thanks for all the quick updates!
I've added 1 last change and then we are good to merge this :tada:

File: services/buildkite/buildkite.tester.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+t.create('buildkite valid pipeline')
+  .get('/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489/master.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'build',
+    value: Joi.equal('failing', 'passing', 'unknown')
+  }));
+
+t.create('buildkite valid pipeline skipping branch')
+  .get('/3826789cf8890b426057e6fe1c4e683bdf04fa24d498885489.json')
+  .expectJSON({ name: 'build', value: 'passing' });
```
Comment (RedSparr0w) [inline services/buildkite/buildkite.tester.js]: Could you add the `Joi.equal('failing', 'passing', 'unknown')` test here also

**Comment by RedSparr0w** [comment]
Thanks, All changes look good.
Merged!
😃

**Comment by raulriera** [comment]
Thanks! is there anything that I need to do to deploy the changes?

**Comment by RedSparr0w** [comment]
Nope, now we just need to wait for it to be deployed to the live server (refer #1538)

-------------
Quoting @paulmelnikow:
>Deploys usually happen every 1–3 weeks. Thaddée, who has limited time on this project, is the only sysadmin. He's working on giving me access to deploy and logs, but doing so is complicated because the hosting account (and maybe the servers too) are shared with other services he runs.

**Pull request merged**